library/src/scala/runtime/LazyVals.scala

```diff
@@ -37,6 +38,41 @@ object LazyVals {
 
   /* ------------- Start of public API ------------- */
 
+  /**
+   * Used to indicate the state of a lazy val that is being
+   * evaluated and of which other threads await the result.
+   */
+  final class Waiting:
+    private var done = false
+    
+    /**
+     * Wakes up waiting threads. Called on completion of the evaluation
+     * of lazy val's right-hand side.
+     */
+    def release(): Unit = synchronized {
+      done = true
+      notifyAll()
+    }
+
+    /**
+     * Awaits the completion of the evaluation of lazy val's right-hand side.
+     */
+    def awaitRelease(): Unit = synchronized {
+      while !done do wait()
+    }
+
+  /**
+   * Used to indicate the state of a lazy val that is currently being
+   * evaluated with no other thread awaiting its result.
+   */
+  object Evaluating
+
+  /**
+   * Used to indicate the state of a lazy val that has been evaluated to
+   * `null`.
+   */
+  object NULL
```

Comment:
Consider using a name that is not full-caps, to follow existing conventions. Perhaps `NullValue`?